File: javascript/all-your-base/all-your-base.js
export function convert(digits, fromBase, toBase) {
    if (!fromBase || fromBase < 2 || !Number.isInteger(fromBase)) {
        throw new Error('Wrong input base');
    }
    if (!toBase || toBase < 2 || !Number.isInteger(toBase)) {
        throw new Error('Wrong output base');
    }
    const hasInvalidDigit = () => digits.some((digit) => digit < 0 || digit >= fromBase);
    if (digits.length == 0 || (digits[0] == 0 && digits.length > 1) || hasInvalidDigit()) {
        throw new Error('Input has wrong format');
    }
    let intermediate = Array.from(digits).reverse()
                                         .map((digit, idx) => digit * (fromBase ** idx))
                                         .reduce((prev, cur) => prev + cur, 0);
    let outDigits = [];
    do {
        outDigits.push(intermediate % toBase);
        intermediate = Math.floor(intermediate / toBase);
    } while (intermediate > 0);
    return outDigits.reverse();
}

File: ruby/all-your-base/all_your_base.rb
module BaseConverter
  def self.convert(input_base, digits, output_base)
    raise ArgumentError, "Input base must be >= 2" unless input_base >= 2
    raise ArgumentError, "Output base must be >= 2" unless output_base >= 2
    raise ArgumentError, "Digits must be >= 0" unless digits.all? { |d| d >= 0 }
    raise ArgumentError, "Digits must be < #{input_base}" unless digits.all? { |d| d < input_base }

    val = digits.reverse.map.with_index { |d, i| d * (input_base ** i) }.sum
    return [0] if val.zero?

    output = []
    while val > 0
      val, digit = val.divmod(output_base)
      output << digit
    end
    output.reverse
  end
end
